devices/flash-flexspi/flashdrv.c

```diff
@@ -0,0 +1,381 @@
+		return -ENXIO;
+	}
+
+	fSz = dev->fspi.slFlashSz[dev->port];
+	fStart = (addr_t)dev->fspi.ahbAddr;
+
+	/* Check if region is located on flash */
+	if ((addr + sz) > fSz) {
+		return -EINVAL;
+	}
+
+	/* Check if flash is mappable to map region */
+	if (fStart <= memaddr && (fStart + fSz) >= (memaddr + memsz)) {
+		*a = fStart;
+		return dev_isMappable;
+	}
+
+	/* Device mode cannot be higher than map mode to copy data */
+	if ((mode & memmode) != mode) {
+		return -EINVAL;
+	}
+
+	/* flash is not mappable to any region in I/O mode */
+	return dev_isNotMappable;
+}
+
+
+static int flashdrv_sync(unsigned int minor)
+{
+	ssize_t res;
+	struct nor_device *dev = minorToDevice(minor);
+	addr_t ofs, pos, sectorAddr = get_sectorAddress(dev, dev->sectorPrevAddr);
+
+	if (dev == NULL || dev->active == 0) {
+		return -ENXIO;
+	}
+
+	/* Initial sector value check, nothing to do */
+	if (dev->sectorPrevAddr == (addr_t)-1) {
+		return EOK;
+	}
+
+	if (dev->sectorPrevAddr == dev->sectorSyncAddr) {
+		return EOK;
+	}
+
+	/* all 'ones' in buffer means sector is erased ... */
+	for (pos = 0; pos < dev->nor->pageSz; ++pos) {
+		if (*(dev->sectorBuf + pos) != NOR_ERASED_STATE) {
+			break;
+		}
+	}
+
+	/* ... than erase may be skipped */
+	if (pos != dev->nor->sectorSz) {
+		res = nor_eraseSector(&dev->fspi, dev->port, sectorAddr, dev->timeout);
+		if (res < 0) {
+			return res;
+		}
+	}
+
+
+	for (ofs = 0; ofs < dev->nor->sectorSz; ofs += dev->nor->pageSz) {
+		/* Just erased NOR page contains all 'ones' */
+		for (pos = 0; pos < dev->nor->pageSz; ++pos) {
+			if (*(dev->sectorBuf + ofs + pos) != NOR_ERASED_STATE) {
+				break;
+			}
+		}
+
+		/* then if buffer is the same skip single page program */
+		if (pos == dev->nor->pageSz) {
+			continue;
+		}
+
+		res = nor_pageProgram(&dev->fspi, dev->port, sectorAddr + ofs, dev->sectorBuf + ofs, dev->nor->pageSz, dev->timeout);
+		if (res < 0) {
+			return res;
+		}
+	}
+
+	hal_cpuInvCache(hal_cpuDCache, sectorAddr, dev->nor->sectorSz);
+
+	dev->sectorSyncAddr = dev->sectorPrevAddr;
+
+	return dev->nor->sectorSz;
+}
+
+
+static ssize_t bufferSync(unsigned int minor, addr_t dstAddr)
+{
+	ssize_t res = 0;
+	struct nor_device *dev = minorToDevice(minor);
+	const int sectorLast = get_sectorIdFromAddress(dev, dev->sectorPrevAddr);
+	const int sectorCurr = get_sectorIdFromAddress(dev, dstAddr);
+
+	if (sectorCurr != sectorLast) {
+		res = flashdrv_sync(minor);
+		if (res < 0) {
+			return res;
+		}
+
+		res = nor_readData(&dev->fspi, dev->port, get_sectorAddress(dev, dstAddr), dev->sectorBuf, dev->nor->sectorSz, dev->timeout);
+		if (res < 0) {
+			return res;
+		}
+
+		dev->sectorPrevAddr = dstAddr;
+	}
+
+	return res;
+}
+
+
+static ssize_t flashdrv_write(unsigned int minor, addr_t dstAddr, const void *data, size_t size)
+{
+	int res;
+	addr_t ofs;
+	const u8 *srcPtr = data;
+	size_t chunkSz, doneBytes;
+	struct nor_device *dev = minorToDevice(minor);
+
+	if (dev == NULL || dev->active == 0)
+		return -ENXIO;
```

Comment:
Inconsistent style

```suggestion
	if (dev == NULL || dev->active == 0) {
		return -ENXIO;
	}
```

Reply:
The mighty power of `git-stash` has revealed itself :)